llvm/lib/Target/AMDGPU/SILowerI1Copies.cpp

```diff
@@ -715,6 +715,14 @@ bool SILowerI1Copies::lowerCopiesToI1() {
       assert(!MI.getOperand(1).getSubReg());
 
       if (!SrcReg.isVirtual() || (!isLaneMaskReg(SrcReg) && !isVreg1(SrcReg))) {
+        if (!SrcReg.isVirtual() &&
+            TII->getRegisterInfo().getRegSizeInBits(SrcReg, *MRI) == 64) {
+          // When calling convention allocates SGPR for i1, for GPUs with
+          // wavefront size 64, i1 return value is put in 64b SGPR.
+          assert(ST->isWave64());
+          continue;
+        }
+
```

Comment:
You don't know this is the context at this point. I don't think you should need to touch this. I assume you encountered a problematic copy here, what exactly does it look like?

Reply:
Immediate after these new lines is an assert():
`assert(TII->getRegisterInfo().getRegSizeInBits(SrcReg, *MRI) == 32);`
which makes sure SrcReg is 32b before lowering the copy to `V_CMP_NE_U32_e64`.

Because, with the callingconv change, the i1 return value is in an SGPR_64 (for wave64 GPUs), these lines were inserted to avoid triggering the assert.

Reply:
But contextually these shouldn't be getting processed. What is the MIR at the point of the assert?

Reply:
The IR instruction callseq_end is always divergent. As a result, the CopyFromReg to copy the return value is also divergent. When the return type is i1, in the machine instruction, the dest reg would be vreg_1. In the above code in SILoweringI1Copies, we avoid lowing this COPY to avoid triggering the assert (code following assert assumes the src is a 32b phys reg).

Reply:
Can you post the actual MIR you are getting? You really should not touch any code in SILowerI1Copies, you do not have the correct context here to know if you're doing something sensible.

callseq_end isn't associated with a value, it's just the SP modification and provides ordering. It's disconnected from the passed register value. In this context, the raw register copy is uniform, but the logical value is divergent. Coming of of the DAG for an incoming value, I would expect you to end up with something like:

```
  %0:sreg_64_exec = COPY $sgpr0_sgpr1 # This ABI copy should exist as-is
  %1:vreg_1 = COPY %0 # SILowerI1Copies is supposed to deal with this 
```

For the outgoing side, I would expect:

```
  %2:vreg_1 = ...
  %3:sreg_64_exec = COPY %2 
  $sgpr0_sgpr1 = COPY %3 # Untouchable ABI copy
```

Reply:
The machine code when reaching SILowerI1Copies is as follows:
```
# Machine code for function test_call_external_i1_func_void: IsSSA, TracksLiveness

bb.0 (%ir-block.0):
  ADJCALLSTACKUP 0, 0, implicit-def dead $scc
  %0:sreg_64 = SI_PC_ADD_REL_OFFSET target-flags(amdgpu-gotprel32-lo) @i1_func_void, target-flags(amdgpu-gotprel32-hi) @i1_func_void, implicit-def dead $scc
  %1:sreg_64_xexec = S_LOAD_DWORDX2_IMM killed %0:sreg_64, 0, 0 :: (dereferenceable invariant load (s64) from got, addrspace 4)
  %2:sgpr_128 = COPY $sgpr0_sgpr1_sgpr2_sgpr3
  $sgpr0_sgpr1_sgpr2_sgpr3 = COPY %2:sgpr_128
  SI_CALL_ISEL killed %1:sreg_64_xexec, @i1_func_void, <regmask $sgpr_null $sgpr_null_hi $src_private_base $src_private_base_hi $src_private_base_lo $src_private_limit $src_private_limit_hi $src_private_limit_lo $src_shared_base $src_shared_base_hi $src_shared_base_lo $src_shared_limit $src_shared_limit_hi $src_shared_limit_lo $sgpr30 $sgpr31 $sgpr32 $sgpr33 $sgpr34 $sgpr35 $sgpr36 $sgpr37 $sgpr38 $sgpr39 $sgpr40 $sgpr41 $sgpr42 $sgpr43 $sgpr44 $sgpr45 $sgpr46 $sgpr47 $sgpr48 and 1194 more...>, implicit $sgpr0_sgpr1_sgpr2_sgpr3, implicit-def $sgpr0_sgpr1
  ADJCALLSTACKDOWN 0, 0, implicit-def dead $scc
  %3:vreg_1 = COPY $sgpr0_sgpr1
  %5:sreg_64_xexec = COPY %3:vreg_1
  %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, 1, %5:sreg_64_xexec, implicit $exec
  %6:sreg_64 = IMPLICIT_DEF
  %7:vreg_64 = COPY %6:sreg_64
  GLOBAL_STORE_BYTE killed %7:vreg_64, killed %4:vgpr_32, 0, 0, implicit $exec :: (volatile store (s8) into `ptr addrspace(1) undef`, addrspace 1)
  SI_RETURN
```
In the COPY of $sgpr0_sgpr1 to %3, the dest reg is vreg_1, and the source is a 64b phys reg. The above code change is to skip this instruction in order to avoid triggering the assert which ensures the src is 32b.

Reply:
```
      MRI->setRegClass(DstReg, IsWave32 ? &AMDGPU::SReg_32RegClass
                                        : &AMDGPU::SReg_64RegClass);
      ...      
      if (!SrcReg.isVirtual() || (!isLaneMaskReg(SrcReg) && !isVreg1(SrcReg))) {
        if (!SrcReg.isVirtual() &&
            TII->getRegisterInfo().getRegSizeInBits(SrcReg, *MRI) == 64) {
          // When calling convention allocates SGPR for i1, for GPUs with
          // wavefront size 64, i1 return value is put in 64b SGPR.
          assert(ST->isWave64());
          continue;
        }

        assert(TII->getRegisterInfo().getRegSizeInBits(SrcReg, *MRI) == 32);
        ...
```
Actually before the IF on line 717 there is a call of `setRegClass` which changes the dest reg
of %3 from vreg_1 to sreg_64. So when line 718 is reached, the insn is already updated to 
`"%3: sreg_64 = COPY $sgpr0_sgpr1"`. The lines added by this patch (718-725) is simply to avoid
triggering the assert on line 726 which assumes that the src phys reg is 32b. For wave64 GPUs, with the calling
conv change introduced by this patch, that assumption is no longer valid.

Reply:
@arsenm Is the code in SILowerI1Copies ok, or do you still have questions?

Reply:
I still think that both:

1. The code already in SILowerI1Copies is nonsense which should be fixed
2. You shouldn't need to touch it

I think the code that's reaching here isn't what you want. You want to have the original copy be the copy to the SReg_64_xexec/SReg_32_xexec, and a copy from there to VReg_1. You have the opposite copy pair here

Reply:
> 1. The code already in SILowerI1Copies is nonsense which should be fixed
Maybe create a separate ticket?

>2. You shouldn't need to touch it
I don't see a better alternative. Do you have any suggestions?

Here is a recap of the full story. I'm listing the .ll file, the init and final DAGs, the instruction selection results and the machine code.
```diff
// .ll file
define i1 @i1_func_void() {
  %val = load i1, ptr addrspace(1) undef
  ret i1 %val
}

define void @test_call_external_i1_func_void() {
  %val = call i1 @i1_func_void()
  store volatile i1 %val, ptr addrspace(1) undef
  ret void
}

// 1. Initial DAG
Initial selection DAG: %bb.0 'test_call_external_i1_func_void:'
SelectionDAG has 17 nodes:
...
  t9: ch,glue = CALL t6, GlobalAddress:i64<ptr @i1_func_void> 0, TargetGlobalAddress:i64<ptr @i1_func_void> 0, Register:v4i32 $sgpr0_sgpr1_sgpr2_sgpr3, RegisterMask:Untyped, t6:1
  t10: ch,glue = callseq_end # D:1 t9, TargetConstant:i64<0>, TargetConstant:i64<0>, t9:1
  - t12: i1,ch,glue = CopyFromReg # D:1 t10, Register:i1 $sgpr0_sgpr1, t10:1
  t14: i64 = Constant<0>
...

+ // Note that CopyFromReg is divergent

// 2. Final DAG
Optimized legalized selection DAG: %bb.0 'test_call_external_i1_func_void:'
SelectionDAG has 21 nodes:
...
  t10: ch,glue = callseq_end # D:1 t9, TargetConstant:i64<0>, TargetConstant:i64<0>, t9:1
  - t12: i1,ch,glue = CopyFromReg # D:1 t10, Register:i1 $sgpr0_sgpr1, t10:1
      t29: i32 = zero_extend # D:1 t12
...

// 3. Instruction selection
===== Instruction selection ends:

Selected selection DAG: %bb.0 'test_call_external_i1_func_void:'
SelectionDAG has 23 nodes:
...
  t10: i1,ch,glue = ADJCALLSTACKDOWN TargetConstant:i64<0>, TargetConstant:i64<0>, t9, t9:1
  - t12: i1,ch,glue = CopyFromReg # D:1 t10:1, Register:i1 $sgpr0_sgpr1, t10:2
      t29: i32 = V_CNDMASK_B32_e64 # D:1 TargetConstant:i32<0>, TargetConstant:i32<0>, TargetConstant:i32<0>, TargetConstant:i32<1>, t12
    t27: ch = GLOBAL_STORE_BYTE<Mem:(volatile store (s8) into `ptr addrspace(1) undef`, addrspace 1)> # D:1 IMPLICIT_DEF:i64, t29, TargetConstant:i16<0>, TargetConstant:i32<0>, t12:1
...

// 4. Machine code at end of ISel
*** MachineFunction at end of ISel ***
# Machine code for function test_call_external_i1_func_void: IsSSA, TracksLiveness

...
  ADJCALLSTACKDOWN 0, 0, implicit-def dead $scc
  - %3:vreg_1 = COPY $sgpr0_sgpr1
  - %5:sreg_64_xexec = COPY %3:vreg_1
  %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, 1, %5:sreg_64_xexec, implicit $exec
  %6:sreg_64 = IMPLICIT_DEF
  %7:vreg_64 = COPY %6:sreg_64
  GLOBAL_STORE_BYTE killed %7:vreg_64, killed %4:vgpr_32, 0, 0, implicit $exec :: (volatile store (s8) into `ptr addrspace(1) undef`, addrspace 1)
...
```

In `InstrEmitter::EmitCopyFromReg()` the CopyFromReg insn transformed to the following:
```
  %3:vreg_1 = COPY $sgpr0_sgpr1
  %5:sreg_64_xexec = COPY %3:vreg_1
```
If I understand you correctly, I think you want the "vreg_1" in %3 to be "sreg_64_xexec". The relevant code of `EmitCopyFromReg()` is as follows:
```
void InstrEmitter::EmitCopyFromReg(...)
  ...
  // Stick to the preferred register classes for legal types.
  if (TLI->isTypeLegal(VT))
    - UseRC = TLI->getRegClassFor(VT, Node->isDivergent());  // UseRC is set to vreg_1 here
...
          const TargetRegisterClass *RC = nullptr;
          if (i + II.getNumDefs() < II.getNumOperands()) {
            RC = TRI->getAllocatableClass(
                TII->getRegClass(II, i + II.getNumDefs(), TRI, *MF));
          }
          - if (!UseRC)  // UseRC is already set, so it's not overwritten here
            UseRC = RC;
...
  if (VRBase) {
    DstRC = MRI->getRegClass(VRBase);
  } else if (UseRC) {
    assert(TRI->isTypeLegalForClass(*UseRC, VT) &&
           "Incompatible phys register def and uses!");
    - DstRC = UseRC;   // DstRC is set to UseRC, which is vreg_1
  } else
    DstRC = SrcRC;
...
```
As you can see, because `UseRC` is set to vreg_1 near the beginning of the function, the final `DstRC` is set to vreg_1 as well.

The code for `SITargetLowering::getRegClassFor()` is as follows:
```cpp
const TargetRegisterClass *SITargetLowering::getRegClassFor(MVT VT, bool isDivergent) const {
  const TargetRegisterClass *RC = TargetLoweringBase::getRegClassFor(VT, false);
  const SIRegisterInfo *TRI = Subtarget->getRegisterInfo();
  if (RC == &AMDGPU::VReg_1RegClass && !isDivergent)  // Returns SReg_64 only for non-divergent SDNode.  
    return Subtarget->getWavefrontSize() == 64 ? &AMDGPU::SReg_64RegClass
                                               : &AMDGPU::SReg_32RegClass;
  if (!TRI->isSGPRClass(RC) && !isDivergent)
    return TRI->getEquivalentSGPRClass(RC);
  else if (TRI->isSGPRClass(RC) && isDivergent)
    return TRI->getEquivalentVGPRClass(RC);

  return RC;
}
```

I don't think changing the behavior of `SITargetLowering::getRegClassFor()` is a good idea. So, I think the current solution, i.e. skipping the instruction, `%3:sreg_64 = COPY $sgpr0_sgpr1` in SILowerI1Copies is preferred.

Pls let me know if you have any suggestions.

Reply:
@arsenm Any suggestions on SILowerI1Copies?

Reply:
The virtual register copies are getting emitted backwards, yes.

You could hack around this in LowerFormalArguments by manually copying to a virtual register with the correct class, and then emitting a copy from that virtual register.

Reply:
This part is for i1 return value though, not i1 arguments.